src/v/v8_engine/script.h

```diff
@@ -73,21 +103,78 @@ class script {
     /// We need to controle execution time for js function
 
     /// \param buffer with data, wich js code can read and edit.
-    void run_internal(ss::temporary_buffer<char>& data);
+    void run_internal(ss::temporary_buffer<char> data);
 
     // Throw c++ exception from v8::TryCatch
+    void throw_exception_from_v8(std::string_view msg);
     void throw_exception_from_v8(
       const v8::TryCatch& try_catch, std::string_view msg);
 
+    ss::future<> add_future_handlers(ss::future<>&& fut) {
+        return fut
+          .handle_exception_type([this](ss::gate_closed_exception&) {
+              throw_exception_from_v8("Executor is stopped");
+          })
+          .handle_exception_type([this](ss::broken_semaphore&) {
+              throw_exception_from_v8("Executor is stopped");
+          });
+    }
+
     struct isolate_deleter {
     public:
         void operator()(v8::Isolate* isolate) const { isolate->Dispose(); }
     };
 
+    // Stop v8 execution if it takes too much time
+    void stop_execution();
+    // Cancel stop execution for reusing isolate
+    void cancel_terminate_execution_for_isolate();
+
     std::unique_ptr<v8::Isolate, isolate_deleter> _isolate;
 
     v8::Global<v8::Context> _context;
     v8::Global<v8::Function> _function;
+
+    // Script timeout
+    std::chrono::milliseconds _timeout_ms;
+
+    // This class implement task for executor. We need to add operator(),
+    // cancel(), on_timeout()
+
+    class task_for_executor {
+    public:
+        task_for_executor(script& script, ss::temporary_buffer<char> data)
+          : _script(script)
+          , _data(std::move(data)) {}
+
+        virtual void operator()() = 0;
+
+        void cancel() { _script.stop_execution(); }
+
+        void on_timeout() { _script.cancel_terminate_execution_for_isolate(); }
+
+    protected:
+        script& _script;
+        ss::temporary_buffer<char> _data;
+    };
+
+    friend class task_for_executor;
+
+    class compile_task : public task_for_executor {
+    public:
+        compile_task(script& script, ss::temporary_buffer<char> data)
+          : task_for_executor(script, std::move(data)) {}
+
+        void operator()() override { _script.compile_script(std::move(_data)); }
```

Comment:
I'm sure we'll go through several versions of this over time, but one thing that you might consider is a model where we allocate all the state that the task will need on the heap and give temporary, full ownership over to the alien thread. this would be in contrast to what is happening here where the alien thread holds a reference to _script.

It's a minor aspect, but it might help us keep state safe from concurrent access as the complexity rises. Here when we execute the remote task on the alien thread, the reactor has to be careful not to touch certain things because that would be unprotected concurrent access.